contracts/cw-rules/src/contract.rs

```diff
@@ -43,33 +44,91 @@ pub fn execute(
         }
+        QueryMsg::GetCW20Balance {
+            cw20_contract,
+            address,
+        } => to_binary(&query_get_cw20_balance(deps, cw20_contract, address)?),
+        QueryMsg::HasBalance {
+            balance,
+            required_balance,
+        } => to_binary(&query_has_balance(balance, required_balance)?),
+        QueryMsg::CheckOwnerOfNFT {
+            address,
+            nft_address,
+            token_id,
+        } => to_binary(&query_check_owner_nft(address, nft_address, token_id)?),
         QueryMsg::CheckProposalReadyToExec {
             dao_address,
             proposal_id,
         } => to_binary(&query_dao_proposal_ready(dao_address, proposal_id)?),
-        QueryMsg::QueryConstruct { rules } => to_binary(&query_construct(deps, env, rules)?),
+        // QueryMsg::QueryConstruct { rules } => to_binary(&query_construct(deps, env, rules)?),
     }
 }
 
 pub fn query_result(_deps: DepsMut, _info: MessageInfo) -> Result<Response, ContractError> {
     Ok(Response::new().add_attribute("method", "query_result"))
 }
 
-// TODO:
-fn query_get_balance(env: Env, address: Addr) -> StdResult<RuleResponse<Option<Binary>>> {
-    Ok((true, None))
+fn query_get_balance(
+    deps: Deps,
+    address: Addr,
+    denom: String,
+) -> StdResult<RuleResponse<Option<Binary>>> {
+    let amount = deps
+        .querier
+        .query_balance(address, denom)
+        .map(|coin| to_binary(&coin.amount))
+        .unwrap()
+        .ok();
+    Ok((true, amount))
 }
 
-// TODO:
-fn query_get_cw20_balance(env: Env, address: Addr) -> StdResult<RuleResponse<Option<Binary>>> {
-    Ok((true, None))
+fn query_get_cw20_balance(
+    deps: Deps,
+    cw20_contract: Addr,
+    address: Addr,
+) -> StdResult<RuleResponse<Option<Binary>>> {
+    let balance: BalanceResponse = deps.querier.query_wasm_smart(
+        cw20_contract,
+        &cw20::Cw20QueryMsg::Balance {
+            address: address.to_string(),
+        },
+    )?;
+    let amount = to_binary(&balance.balance).ok();
+    Ok((true, amount))
+}
+
+fn query_has_balance(
+    balance: Balance,
+    required_balance: Balance,
+) -> StdResult<RuleResponse<Option<Binary>>> {
+    let res = match balance {
+        Balance::Native(native) => match required_balance {
+            Balance::Native(native_required) => {
+                let mut res_native = true;
+                for coin in native_required.0 {
+                    if !has_coins(&native.0, &coin) {
+                        res_native = false;
+                        break;
+                    }
+                }
+                res_native
+            }
+            Balance::Cw20(_) => false,
+        },
+        Balance::Cw20(cw20) => match required_balance {
+            Balance::Native(_) => false,
+            Balance::Cw20(cw20_required) => {
+                cw20.address == cw20_required.address && cw20.amount >= cw20_required.amount
+            }
+        },
+    };
```

Comment:
```suggestion
let res = match (balance, required_balance) {
        (Balance::Native(current), Balance::Native(expected)) => {
            expected.0.iter().all(|c| has_coins(&current.0, c))
        }
        (Balance::Cw20(current_cw20), Balance::Cw20(expected_cw20)) => {
            current_cw20.address == expected_cw20.address
                && current_cw20.amount >= expected_cw20.amount
        }
        _ => false,
    };
```
little bit more compact